### PR TITLE
BLOCKCHAIN-221 FE - fix preimage already exists issues

### PR DESCRIPTION
**Pull request by MichalLustykNeti**

### CODE REVIEW

File: src/api/nodeRpcCall.js [hunk 1]
```diff
@@ -1845,14 +1834,19 @@ const congressProposeLegislationReferendum = async (
     referendumPropose;
 
   const threshold = await congressMajorityThreshold();
+  const existingPreimages = await api.query.preimage.preimageFor([addLaw.hash, addLaw.encodedLength])
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:1827]:
```suggestion
  const existingPreimage = await api.query.preimage.preimageFor([addLaw.hash, addLaw.encodedLength])
```

But TBH I would even reduce it to just `preimage`.

File: src/api/nodeRpcCall.js [hunk 2]
```diff
@@ -774,37 +766,24 @@ const submitProposal = async (walletAddress, values, callback) => {
...
-        callback({isError: true, details: error.toString()});
-      });
-    }
-  }).catch((error) => {
-    // eslint-disable-next-line no-console
-    console.error(':( transaction failed', error);
-    callback({isError: true, details: error.toString()});
-  });
+
+  const existingPreimages = await api.query.preimage.preimageFor([extrinsicMethod.hash, extrinsicMethod.encodedLength])
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:778]:
```suggestion
  const existingPreimage = await api.query.preimage.preimageFor([extrinsicMethod.hash, extrinsicMethod.encodedLength])
```
But TBH I would reduce it to just `preimage`

File: src/api/nodeRpcCall.js [hunk 3]
```diff
@@ -774,37 +766,24 @@ const submitProposal = async (walletAddress, values, callback) => {
     additionalMetadata: {},
     proposerAddress: walletAddress,
   });
-  const legislationIndex = centralizedMetadata.data.id;
-  const hash = await getProposalHash(values, legislationIndex);
-  const notePreimageTx = api.tx.preimage.notePreimage(hash.extrinsicEncoded);
+  const extrinsicMethod = api.tx.liberlandLegislation.addLaw(
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:769]:
```suggestion
  const addLaw = api.tx.liberlandLegislation.addLaw(
```
`extrinsic` is the whole transaction. And is a bit too generic, given that we also operate on `democracy.propose*`, `preimage` and `utility.batchAll` calls

File: src/api/nodeRpcCall.js [hunk 4]
```diff
@@ -774,37 +766,24 @@ const submitProposal = async (walletAddress, values, callback) => {
...
-    // eslint-disable-next-line no-console
-    console.error(':( transaction failed', error);
-    callback({isError: true, details: error.toString()});
-  });
+
+  const existingPreimages = await api.query.preimage.preimageFor([extrinsicMethod.hash, extrinsicMethod.encodedLength])
+  const extrinsic = existingPreimages.isNone
+  ? api.tx.utility.batchAll([
+      api.tx.preimage.notePreimage(extrinsicMethod.toHex()),
+      proposeCall({ Legacy: extrinsicMethod.hash }, minDeposit),
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:782]: Lets upgrade from `Legacy` to `Lookup`.

File: src/api/nodeRpcCall.js [hunk 5]
```diff
@@ -774,37 +766,24 @@ const submitProposal = async (walletAddress, values, callback) => {
...
+    ])
+  : proposeCall({ Legacy: extrinsicMethod.hash }, minDeposit);
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:784]: Ditto.

File: src/api/nodeRpcCall.js [hunk 6]
```diff
@@ -1884,10 +1868,16 @@ const congressProposeRepealLegislation = async (
 
   const congressmen = await api.query.council.members();
   const threshold = Math.trunc(congressmen.length/2) + 1;
-  const extrinsic = api.tx.utility.batchAll([
-    api.tx.preimage.notePreimage(repealLaw.toHex()),
-    api.tx.council.propose(threshold, proposal, proposal.length),
-  ])
+
+  const existingPreimages = await api.query.preimage.preimageFor([repealLaw.hash, repealLaw.encodedLength])
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:1872]:
```suggestion
  const existingPreimage = await api.query.preimage.preimageFor([repealLaw.hash, repealLaw.encodedLength])
```
But TBH I would reduce it to just preimage

File: src/redux/sagas/base.js [hunk 7]
```diff
@@ -19,3 +19,7 @@ function errorHandler(onFailure, worker) {
 export function* blockchainWatcher({ call, failure }, worker) {
   yield takeLatest(call, errorHandler(failure, worker));
 }
+
+export function* democracyWatcher({ call, failure }, worker) {
```
Comment (kacperzuk-neti) [inline src/redux/sagas/base.js:23]: Nooo, just use `blockchainWatcher` everywhere.

File: src/api/nodeRpcCall.js [hunk 8]
```diff
@@ -752,16 +752,8 @@ const voteOnReferendum = async (walletAddress, referendumIndex, voteType, callba
...
 };
 
-const getProposalHash = async (values, legislationIndex) => {
-  const api = await getApi();
-  const extrinsicEncoded = api.tx.liberlandLegislation.addLaw(parseInt(values.legislationTier), legislationIndex, values.legislationContent).method.toHex();
-  const hash = { encodedHash: blake2AsHex(extrinsicEncoded), extrinsicEncoded };
-  return hash;
-};
-
 const submitProposal = async (walletAddress, values, callback) => {
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:755]: `callback` is now unused.

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]

**Review by kacperzuk-neti — APPROVED** [review]

**Comment by DorianSternVukotic** [comment]
Why

**Comment by kacperzuk-neti** [comment]
Other than eslint fixes and refactoring to use `submitExtrinsic`, the main set of chances is this:

```
  const existingPreimage = await api.query.preimage.preimageFor([repealLaw.hash, repealLaw.encodedLength])

  const extrinsic = existingPreimages.isNone
  ? api.tx.utility.batchAll([
      api.tx.preimage.notePreimage(repealLaw.toHex()),
      api.tx.council.propose(threshold, proposal, proposal.length),
    ])
  : api.tx.council.propose(threshold, proposal, proposal.length);
  ```

If there already is a preimage for the exact call you want to make, `notePreimage` will fail saying that you can't have duplicates. So this PR detects that and skips `notePreimage` if we already have the preimage.
I assume this will rarely matter in production, unless Senate/Congress use their cancellation rights. It's mainly annoying during testing.

File: src/api/nodeRpcCall.js [hunk 9]
```diff
@@ -1845,14 +1824,19 @@ const congressProposeLegislationReferendum = async (
     referendumPropose;
 
   const threshold = await congressMajorityThreshold();
+  const existingPreimage = await api.query.preimage.preimageFor([addLaw.hash, addLaw.encodedLength])
+
+  const extrinsic = existingPreimages.isNone
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:1829]:
```suggestion
  const extrinsic = existingPreimage.isNone
```

File: src/api/nodeRpcCall.js [hunk 10]
```diff
@@ -1884,10 +1868,16 @@ const congressProposeRepealLegislation = async (
...
   const congressmen = await api.query.council.members();
   const threshold = Math.trunc(congressmen.length/2) + 1;
-  const extrinsic = api.tx.utility.batchAll([
-    api.tx.preimage.notePreimage(repealLaw.toHex()),
-    api.tx.council.propose(threshold, proposal, proposal.length),
-  ])
+
+  const existingPreimage = await api.query.preimage.preimageFor([repealLaw.hash, repealLaw.encodedLength])
+  
+  const extrinsic = existingPreimages.isNone
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:1874]:
```suggestion
  const extrinsic = existingPreimage.isNone
```

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]

**Review by kacperzuk-neti — APPROVED** [review]

**Comment by DorianSternVukotic** [comment]
Ok, solve conflicts and merge when ready

**Pull request merged**